tools/cluster_mgr.py

```diff
@@ -153,11 +149,180 @@ def main():
+
+    client = redis.Redis(decode_responses=True,
+                         host="localhost", port=args.first_port)
+    existing = client.execute_command("cluster", "shards")
+    config = []
+    for shard_list in existing:
+        shard = list_to_dict(shard_list)
+        config.append({
+            "slot_ranges": build_slots(shard["slots"]),
+            "master": build_node(shard["nodes"][0]),
+            "replicas": [build_node(replica) for replica in shard["nodes"][1::]]
+        })
+    return config
+
+
+def move(args):
+    config = build_config_from_existing(args)
+
+    new_owner = None
+    for shard in config:
+        if shard["master"]["port"] == args.target_port:
+            new_owner = shard
+            break
+    else:
+        print(
+            f"Can't find master with port {args.target_port} (hint: use flag --target_port).")
+        exit(-1)
+
+    def remove_slot(slot, from_range, from_shard):
+        if from_range["start"] == slot:
+            from_range["start"] += 1
+            if from_range["start"] > from_range["end"]:
+                from_shard["slot_ranges"].remove(from_range)
+        elif from_range["end"] == slot:
+            from_range["end"] -= 1
+            if from_range["start"] > from_range["end"]:
+                from_shard["slot_ranges"].remove(from_range)
+        else:
+            assert slot > from_range["start"] and slot < from_range["end"], f'{slot} {from_range["start"]} {from_range["end"]}'
+            from_shard["slot_ranges"].append(
+                {"start": slot + 1, "end": from_range["end"]})
+            from_range["end"] = slot - 1
+
+    def add_slot(slot, to_shard):
+        for slot_range in to_shard["slot_ranges"]:
+            if slot == slot_range["start"] - 1:
+                slot_range["start"] -= 1
+                return
+            if slot == slot_range["end"] + 1:
+                slot_range["end"] += 1
+                return
+        to_shard["slot_ranges"].append({"start": slot, "end": slot})
+
+    def find_slot(slot, config):
+        for shard in config:
+            if shard == new_owner:
+                continue
+            for slot_range in shard["slot_ranges"]:
+                if slot >= slot_range["start"] and slot <= slot_range["end"]:
+                    return shard, slot_range
+        return None, None
+
+    def pack(slot_ranges):
+        for slot_range in slot_ranges:
+            for other_slot_range in slot_ranges:
+                if slot_range["end"] + 1 == other_slot_range["start"]:
+                    slot_range["end"] = other_slot_range["end"]
+                    slot_ranges.remove(other_slot_range)
+                    break
+
+    for slot in range(args.slot_start, args.slot_end+1):
+        shard, slot_range = find_slot(slot, config)
+        if shard == None:
+            continue
+        if shard == new_owner:
+            continue
+        remove_slot(slot, slot_range, shard)
+        add_slot(slot, new_owner)
```

Comment:
I never saw them being sorted anywhere, so you assume that with a bad config they can be in a random order? And you also insert them one by one. This can add up to quite a few operations given that there are 16k slots. Like a couple of hundred thousands. Is it true for the worst case? I mean it can add up to a second or such

Reply:
It's actually not bad config to have them unsorted.
And yeah, I don't care if the script takes a few seconds to run :)